### DISCUSSION TITLE
Option to write "First Load Size" data to file

### DISCUSSION THREAD

**Discussion by merrywhether** [vercel/next.js#23977] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

We'd previously used the [bundlesize](https://github.com/siddharthkp/bundlesize) package in CI in order to monitor the growth of our application and prevent accidental imports of test libs, etc. With the switch to the new chunking strategy, the hash-based names of the new chunks breaks this approach (we previously just watched `chunks/commons.*.js`). But the new "First Load Size" data could easily replace this functionality, if it were easily accessible (I'm not super keen on redirecting and parsing stdout).

## Describe the solution you'd like

Would it be possible to add a configuration option to next.config that enabled also writing the size data to a json file such that it could be then analyzed by other tools?

## Describe alternatives you've considered

Capturing and parsing the stdout of a yarn build command.

**Comment by Timer** [comment]
Hey! We probably won't a file dedicated to file sizes as it's too specialized.

You can self-compute this using `.next/build-manifest.json`, though!

**Comment by timneutkens** [comment]
I'd like to keep this open, it's an interesting idea.

**Comment by merrywhether** [comment]
Is this something you'd take a PR for, if I was able to get approval from my company? The addition actually looks pretty straightforward in `next/build/index`, since the data is all there in the `pageInfos` structure.

Thinking about it more, seems like the best approach could actually be writing out that entire data structure such that people can programmatically interact with the data in it, since people might want to assert static renderability or other things as well. `persistPageStatsToFile` or something could just add `pageStats.json` to `.next/`?

**Comment by timneutkens** [comment]
We’d have to figure out the exact semantics, we’re currently focusing on other things (SSG, rewrites/redirects) but eventually we’ll add this when we know the right semantics needed to build these kind of tools.

**Comment by jtthackery** [comment]
~I was just looking at the @next/bundle-analyzer package to do something like this, would it be possible for that to take in some optional config for output location and type (i.e. json or text instead of html)?  It seems like the underlying `webpack-bundle-analyzer` package can take those options.~

Heh, just saw #11066 and the comment there, nevermind then!

**Comment by bobaaaaa** [comment]
This should be now possible with nextjs 9.5. See: https://nextjs.org/blog/next-9-5#persistent-caching-for-page-bundles

Fixed by #13937

**Comment by merrywhether** [comment]
Stable hashes doesn't address the issue of trying to reason about the size of a page, since we'd still have to add up the bundle parts included in each page. Something like `bundlesize` could already work with the regex-addressable page files; the problem is that each page needs an "arbitrary" collection of intermediate JS files. That data is perfectly contained in the `pageInfos` data structure but not exposed post build so it can't be used for monitoring.

I know bundle size isn't everything when it comes to speed of delivery, but it's a pretty useful first line of defense against someone accidentally including something like `faker` outside of test files, for instance.

**Comment by izi-p** [comment]
Hey all, 

I wrote a tiny script that computes "first load sizes" of your next.js pages, and store them in a json file.

https://github.com/izi-p/next-compute-first-load

Feel free to use it  🥰

Thank you
